### PR TITLE
Convex Destination Connector (internal review)

### PR DESCRIPTION
**Pull request by emmaling27**
## What
I work at Convex, where we're building a platform for building apps. We already have a Convex source connector, and we'd like to add a Convex destination connector.

This PR adds a new destination connector to import data into Convex (https://convex.dev/).

## How
1. Generate a new connector from the [Python CDK tutorial](https://docs.airbyte.com/connector-development/tutorials/cdk-tutorial-python-http/getting-started)
2. Created new endpoints in the Convex API: `/clear_tables`, `/airbyte_ingress`
3. Updated the new connector to use the APIs, until `check` and `write` work
4. Wrote unit tests with mock responses for Convex API.

## Recommended reading order
1. `destination_convex/spec.json`
2. `destination_convex/destination.py`
3. `destination_convex/writer.py`
4. `destination_convex/client.py`
5. `integration_tests/integration_test.py`

## 🚨 User Impact 🚨
New destination connector in release stage "alpha".

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [X] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

Unit test output:
```
=========================================================================================================== test session starts ===========================================================================================================
platform darwin -- Python 3.9.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /Users/emmaling/Projects/airbyte/airbyte-integrations/connectors/destination-convex/.venv/bin/python
cachedir: .pytest_cache
rootdir: /Users/emmaling/Projects/airbyte, configfile: pytest.ini
collected 2 items                                                                                                                                                                                                                         

unit_tests/unit_test.py::test_check PASSED
unit_tests/unit_test.py::test_write PASSED

============================================================================================================ 2 passed in 0.22s ===========================================================================================================
```
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-convex/destination_convex/client.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, List, Mapping
+
+import requests
+from destination_convex.config import ConvexConfig
+
+
+class ConvexClient:
```
Comment (ldanilek) [inline airbyte-integrations/connectors/destination-convex/destination_convex/client.py:11]: i don't think airbyte runs python typechecking. the way i did this was copying the client.py file into the convex repo / smoketest directory and making sure the types work there.

File: airbyte-integrations/connectors/destination-convex/destination_convex/destination.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            if message.type == Type.STATE:
+                # Emitting a state message indicates that all records which came before it have been written to the destination. So we flush
+                # the queue to ensure writes happen, then output the state message to indicate it's safe to checkpoint state
+                writer.flush()
+                yield message
+            elif message.type == Type.RECORD and message.record is not None:
+                msg = message.record.dict()
+                writer.queue_write_operation(msg)
+            else:
+                # ignore other message types for now
```
Comment (ldanilek) [inline airbyte-integrations/connectors/destination-convex/destination_convex/destination.py:77]: what other message types are there and which ones might we want to handle in the future?

File: docs/integrations/destinations/convex.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+- Convex Account
+- Convex Project
+- Deploy key
+
+### Setup guide
+
+Airbyte integration is available to all Convex developers.
+
+On the [Convex dashboard](https://dashboard.convex.dev/), navigate to the project that you want to sync.
+Note only "Production" deployments should be synced.
```
Comment (ldanilek) [inline docs/integrations/destinations/convex.md]: there may be use-cases to sync data from a test/staging data store into dev deployments. and i think convex would support that

**Review by ldanilek — APPROVED** [review]
looks great!

File: airbyte-integrations/connectors/destination-convex/destination_convex/destination.py (hunk 2 above)
Comment (emmaling27) [reply airbyte-integrations/connectors/destination-convex/destination_convex/destination.py:77]: oh this is copy + pasted from their template. the other types are:
```
    LOG = "LOG"
    SPEC = "SPEC"
    CONNECTION_STATUS = "CONNECTION_STATUS"
    CATALOG = "CATALOG"
    TRACE = "TRACE"
    CONTROL = "CONTROL"
```
I don't think the destination needs to do anything with these for the destination connector <-> convex interaction. I don't think the airbyte <-> destination connector interaction needs them either but we'll see with their review.

File: docs/integrations/destinations/convex.md (hunk 3 above)
Comment (emmaling27) [reply docs/integrations/destinations/convex.md]: good point, I'll update this

**Pull request merged**